### PR TITLE
TS Insert NULL support

### PR DESCRIPTION
**Pull request by paegun**
is depended upon by:
- https://github.com/basho/riak_kv/pull/1507

does:
- for Riak TS insert support for NULL values, implicitly and explicitly, changed riak_ql_ddl:insert_sql_columns/3 to allow the ddl module to flesh out fields and values not specified in the query.

While the column reordering may look like a bunch of avoidable work (reduced dev- and run-time), this feature is already published as supported w/i Riak TS documentation, but was not actually supported prior to this work, see:
http://docs.basho.com/riak/ts/1.4.0/using/writingdata/#adding-data-via-sql

> As with standard SQL, if all of the field names are not provided before the VALUES keyword, the other fields are assumed to be null.
> 
> The fields can be in any order, but the field name and the values must match up. Without the VALUES keyword, all fields must be present in the same order as the schema definition.

### CODE REVIEW

File: src/riak_ql_ddl.erl [hunk 1]
```diff
@@ -1174,6 +1344,14 @@ is_query_valid_no_subexpressions_1_test() ->
                                   "AND myfamily = 'bob' ")
       ).
 
+is_insert_valid_null_test() ->
```
Comment (macintux) [inline src/riak_ql_ddl.erl:1354]: I think this test (or a similar one) needs to check the resulting record; any unquoted strings are being interpreted as strings.

File: src/riak_ql_lexer.xrl [hunk 2]
```diff
@@ -177,19 +176,31 @@ lex(String) ->
...
-    list_to_binary(DeDoubledInternalQuotes).
-
-strip_quoted(QuotedString) ->
-    % if there are unicode characters in the string, throw an error
-    [error(unicode_in_quotes) || U <- QuotedString, U > 127],
-
-    StrippedOutsideQuotes = accurate_strip(QuotedString, $"),
-    re:replace(StrippedOutsideQuotes, "\"\"", "\"", [global, {return, binary}]).
+    Literal1 = case hd(Literal) of
+        $' -> accurate_strip(Literal, $');
```
Comment (hmmr) [inline src/riak_ql_lexer.xrl:194]: Because `CHARACTER_LITERAL` is now obligated to be single-quoted, you can simply and unconditionally drop the first and last characters of `Literal` here (and remove `accurate_strip/1`).

File: src/riak_ql_lexer.xrl [hunk 3]
```diff
@@ -177,19 +176,31 @@ lex(String) ->
...
+        $" ->
+            [error(unicode_in_quotes) || U <- Literal, U > 127],
```
Comment (hmmr) [inline src/riak_ql_lexer.xrl:196]: I suggest to only report the first occurrence of unicode characters, rather than each of them (which can be indefinitely many).

Comment (paegun) [reply src/riak_ql_lexer.xrl:196]: This list comprehension errors on first occurrence, so does what you're suggesting :)

File: src/riak_ql_lexer.xrl (hunk 2 above)
Comment (paegun) [reply src/riak_ql_lexer.xrl:194]: Agreed, but for DRY sake, since the same is true for a quoted identifier, stripping the first and last quote character should be in a function, so accurate_strip/1 .

**Comment by hmmr** [comment]
+1

File: src/riak_ql_ddl.erl [hunk 4]
```diff
@@ -546,24 +546,98 @@ make_empty_insert_row(Mod) ->
...
+%% If the INSERT command specifies only a partial list of columns, expand the
+%% list of columns to include those known by the DDL, ensuring that the values
+%% provided and those omitted align with the columns as provided by the caller.
+-spec insert_sql_columns(module(), [field_identifier()], [data_value()]) ->
+    {[field_identifier()],[[data_value()]]}.
+insert_sql_columns(Mod, Fields, Values) when is_atom(Mod) ->
+    FInMod = [ F || {[F], _I} <- Mod:get_field_positions() ],
+    case any_unknown_column(Fields, FInMod, false) of
+        true ->
+            %% will fail downstream
```
Comment (hmmr) [inline src/riak_ql_ddl.erl:568]: You may be right in letting this error be handled in `riak_kv_qry:do_insert/1` (eventually in `:lookup_field_positions/2`). Detecting and reporting it here will break existing tests (which in itself is not so much of a problem), but more importantly, will render the error handling code in riak_kv_qry effectively dead -- which in turn will surely raise the question of removing it. I would certainly prefer to stick a comment here about relying on error handling code downstream and be done with it, just as you do here.

File: src/riak_ql_ddl.erl [hunk 5]
```diff
@@ -546,24 +546,98 @@ make_empty_insert_row(Mod) ->
...
+        false -> lists:sublist(Values, length(FInQuery));
+        _ ->
+            case length(Values) =:= length(FInQuery) of
+                true -> Values;
+                _ ->
+                    Values ++
+                    [ {null, ?SQL_NULL} || _I <- lists:seq(length(Values) + 1, length(FInQuery)) ]
+            end
+    end,
+    lists:foldr(fun ({Type, Value}, Acc) ->
```
Comment (hmmr) [inline src/riak_ql_ddl.erl]: For curiosity's sake, why `foldr` and not `foldl`?

File: src/riak_ql_ddl.erl [hunk 6]
```diff
@@ -1017,6 +1095,100 @@ is_insert_valid_test_helper(Table_name, Table_def, Insert) ->
...
+    ValuesIn = [[{varchar, <<"family1">>},
+            {varchar, <<"series1">>},
+            {timestamp, 1000},
+            {varchar, <<"sunny">>},
+            {double, 37.0}]],
+    {FieldsOut, ValuesOut } = insert_sql_columns(Mod, FieldsIn, ValuesIn),
+    ?assertEqual(FInMod, FieldsOut),
+    ?assertEqual(ValuesIn, ValuesOut).
+
+insert_sql_columns_pk_fields_test() ->
```
Comment (hmmr) [inline src/riak_ql_ddl.erl:1111]: This test seems redundant. The fields constituting primary key are nothing special in themselves with respect to how their null/non-null attribute should be treated in parsing an INSERT query.

File: src/riak_ql_ddl.erl [hunk 7]
```diff
@@ -1017,6 +1095,100 @@ is_insert_valid_test_helper(Table_name, Table_def, Insert) ->
...
+    ?assertEqual(ValuesIn, ValuesOut).
+
+%% Shuffle two equal length lists equivalently
+shuffle2(L0, L1) ->
+    shuffle2(list_to_tuple(L0), list_to_tuple(L1), length(L0)).
+
+shuffle2(L0, L1, 0) ->
+    {tuple_to_list(L0), tuple_to_list(L1)};
+shuffle2(L0, L1, Len) ->
+    Rand = random:uniform(Len),
```
Comment (hmmr) [inline src/riak_ql_ddl.erl]: Rather than introducing nondeterministic behaviour in tests, I would suggest running your `insert_sql_columns_shuffled_fields_test` through a battery of predefined combinations of fields vs values (or, go the whole hog and write some EQC generators :)

File: src/riak_ql_lexer.xrl (hunk 3 above)
Comment (hmmr) [reply src/riak_ql_lexer.xrl:196]: Those non-local exits. I see.

File: test/parser_insert_tests.erl [hunk 8]
```diff
@@ -72,3 +72,34 @@ insert_sint64_test() ->
...
+insert_identifier_test() ->
+    Insert_sql =
+        "INSERT INTO mytab (col) VALUES (john)",
+    ?assertEqual(
+        {insert,[{table,<<"mytab">>},
+                                    {fields,[{identifier,[<<"col">>]}]},
+                                    {values,[[{identifier,<<"john">>}]]}]},
+        riak_ql_parser:ql_parse(riak_ql_lexer:get_tokens(Insert_sql))
+    ).
+
```
Comment (hmmr) [inline test/parser_insert_tests.erl:115]: How does the parser cope with a null literal deceptively quoted, e.g., `"INSERT INTO mytab VALUES ('null')`"?

Comment (paegun) [reply test/parser_insert_tests.erl:115]: `'null'` is a character literal which has the value `['n','u','l','l']` which is NOT `NULL` which has the value `[]`. The parser treats:
- unquoted NULL (and all case-insensitive variants) as NULL
- quoted 'null' (and all quoted content for that matter) as a character literal

Comment (hmmr) [reply test/parser_insert_tests.erl:115]: Yes, but.. I meant to say a test where `'null`' gets parsed as `{binary, <<"null">>}` would be appropriate :)

File: src/riak_ql_ddl.erl (hunk 7 above)
Comment (paegun) [reply src/riak_ql_ddl.erl]: All but one case of the shuffle covers the case of the insert statement handling columns specified out of order. Hence, going whole hog is over coverage of the feature. The one case that fails to provide coverage is when the shuffle results in the column order being mapped back to the exact same column order.

I considered swapping the first and last columns to achieve this result, but felt that the odds of an exact order match were low enough to have this more random method of coverage provide the coverage.

I also considered using function guards:
- when the resulting list is empty (entering the recursion), start the shuffle
- when the resulting list matches the input list, restart the shuffle
- when the resulting list matches the length of the input list (but not the exact content), return

This seems like overkill.

File: src/riak_ql_ddl.erl (hunk 6 above)
Comment (paegun) [reply src/riak_ql_ddl.erl:1111]: Only specifying primary key fields is the lower-boundary. A query must specify at minimum the primary keys. Other tests that provide partial field specification do not test at this boundary. Generally, testing should test at the boundaries as well as within the boundary.

File: test/parser_insert_tests.erl (hunk 8 above)
Comment (paegun) [reply test/parser_insert_tests.erl:115]: Understood. My thinking was that covering 'null' was overkill as that's just a varchar like any varchar, BUT you are right, a developer could make the mistake of thinking 'null' is NULL. I'll add a test for 'null'.

File: src/riak_ql_ddl.erl (hunk 1 above)
Comment (paegun) [reply src/riak_ql_ddl.erl:1354]: insert_identifier_test covers the input side of this. w/i riak_kv, erroring on an identifier w/i the values is handled and tested.

**Comment by hmmr** [comment]
+1

I can see how this has no ill-effects in this PR, but I still am convinced that `random` functions have no place in any tests.

File: src/riak_ql_ddl.erl (hunk 7 above)
Comment (macintux) [reply src/riak_ql_ddl.erl]: I too am a bit uneasy with the approach.

I vote for either EQC or generating several inputs (randomly is fine) for static comparison. Since EQC appears to be overkill per your analysis, the latter should be sufficient.

**Comment by hmmr** [comment]
+1

File: src/riak_ql_ddl.erl [hunk 9]
```diff
@@ -974,10 +1048,14 @@ parsed_sql_to_query(Proplist) ->
 
 %% is_query_valid expects a 3-tuple: table name, fields, values
 parsed_sql_to_insert(Mod, Proplist) ->
+    Table = proplists:get_value(table, Proplist, <<>>),
```
Comment (hazen) [inline src/riak_ql_ddl.erl:1051]: Much clearer and cleaner

File: src/riak_ql_ddl.erl [hunk 10]
```diff
@@ -1229,6 +1413,15 @@ is_insert_valid_invalid_column_1_test() ->
...
             "('hazen', 'world', 'cloudy', 15)")).
 
+is_insert_valid_multiple_rows_test() ->
+    ?assertEqual(
+        true,
+        is_insert_valid_test_helper("mytab", ?LARGE_TABLE_DEF,
+            "INSERT INTO mytab VALUES"
+            "('hazen', 'world', 70, 'sunny', 45.0),"
+            "('hazen', 'world', 71, 'sunny', 44.0),"
+            "('hazen', 'world', 72, 'sunny', 46.0)")).
```
Comment (hazen) [inline src/riak_ql_ddl.erl:1423]: Excellent choice of test data

File: src/riak_ql_lexer.xrl [hunk 11]
```diff
@@ -188,18 +190,27 @@ clean_up_hex([$0,$x|Hex]) ->
...
-                                         "''", "'",
-                                         [global, {return, list}]),
-    list_to_binary(DeDoubledInternalQuotes).
-
-strip_quoted(QuotedString) ->
-    % if there are unicode characters in the string, throw an error
-    [error(unicode_in_quotes) || U <- QuotedString, U > 127],
-
-    StrippedOutsideQuotes = accurate_strip(QuotedString, $"),
-    re:replace(StrippedOutsideQuotes, "\"\"", "\"", [global, {return, binary}]).
```
Comment (hazen) [inline src/riak_ql_lexer.xrl:202]: Guessing you looked at the `re:replace` performance which is why you wrote your own?

**Comment by hazen** [comment]
:shipit:

**Review by hazen — COMMENTED** [review]
:+1:

**Comment by hazen** [comment]
+1
